### ISSUE TITLE
space_before_paren has a false-positive

### ISSUE THREAD

**Issue by kgrubb** [vala-lang/vala-lint#104]
The regex for `space_before_paren` will catch  pointer dereference operators and expect a space, which is probably not what we want.

Example situation: https://github.com/elementary/switchboard-plug-wacom/pull/7

What we'd probably expect:
```vala
// pointer dereference, we don't really want a space here
X.Atom device_type = *((X.Atom*)data);
// multiplication operator, we do want a space here
var test = 2 * (3 + 1);
```
This is fairly tricky since * can be the multiplication operator which should include a space. Here's some examples with the current regex:
![Screenshot from 2019-09-17 22-21-51](https://user-images.githubusercontent.com/8471701/65102597-92f0f780-d999-11e9-8f99-92bb481d9734.png)

**Comment by pantor** [comment]
This case is really hard to find with a regex-based approach. Instead, the `space-before-paren` check should make use of the AST for that.
